src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,134 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final Object[] EMPTY_ELEMENT_DATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size = 0;
+
+    public ArrayList() {
+        this.elementData = EMPTY_ELEMENT_DATA;
+    }
+
+    private void quickRemove(Object[] elements, int index) {
+        if ((size - 1) >= index) {
+            System.arraycopy(elements, index + 1, elements, index, (size - 1) - index);
+        }
+        elements[size -= 1] = null;
+    }
+
+    private void checkIndex(int index) {
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index is out of bound of array");
+        }
+    }
+
+    private Object[] grow() {
+        return grow(size + 1);
+    }
+
+    private Object[] grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != EMPTY_ELEMENT_DATA) {
+            int newCapacity = oldCapacity >> 1;
+            Object[] copyOfElementData = elementData;
+            elementData = new Object[Math.max(newCapacity,minCapacity)];
+            System.arraycopy(copyOfElementData,0, elementData, 0, oldCapacity);
+            return elementData;
+        }
+        return elementData = new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+    }
+
+    private void add(T e,Object[] element,int size) {
+        if (size == element.length) {
+            element = grow();
+        }
+        element[size] = e;
+        this.size = size + 1;
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, elementData, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        add(value, elementData, size);
+        checkIndex(index);
```

Comment:
The `checkIndex` method should be called before any other logic in the `add(T value, int index)` method to prevent any changes if the index is out of bounds.